keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1838 @@
+a reason type. It is also possible some of the more general api errors have error 
+messages that contain context not found in OpenAPI Error.
+
+##### Mitigation: Fallback to CEL
+
+For cases where the potentially lost error message context must be retained,
+we can instead encode the rule as a CEL rule which allows for arbitrary error
+strings.
+
+##### Mitigation: Explore `x-kubernetes` extension for custom error messages
+
+An option to consider to all custom error messages to schema-based value 
+validations if we see that we really need it is to introduce a new extension:
+`x-kubernetes-validation-errors` which maps schema validation fields to
+a custom error message for failing that field's validation:
+
+```yaml
+minLength: 0
+x-kubernetes-validation-errors:
+  minLength:
+    message: "this custom error message shows when violating minLength rather than default message"
+    reason: Invalid
+```
+
+#### Risk: Added latency to API request handling.
+
+The design decision to declaratively validate API versions has performance
+implications.
+
+Today, the apiserver's request handler decodes incoming versioned API requests
+to the versioned type and then immediately performs an "unsafe" conversion to
+the internal type when possible. All subsequent request processing uses the
+internal type as the "hub" type. Validation is written for the internal
+type and so are per-resource strategies.
+
+With this change, the internal type will no longer be responsible for
+validation.
+
+If we were to convert from the internal back to the version of the API request
+for validation, we would introduce one additional conversion. If we make this
+an "unsafe" conversion, than is will be low cost for the vast majority of requests.
+
+We will benchmark this approach and plan to use it for alpha.
+
+Long term, we could do better:
+
+Since the internal type will no longer be used for validation, it becomes a lot
+less important. It is still important to have a hub type. But why not pick one
+of the versioned types to be the hub type? The vast majority of APIs only have
+one version anyway. The obvious candidate version to choose for the hub version
+would the perferred storage version.
+
+Switching to a versioned type for the hub type would have a few implications:
+
+- We would eliminate the need for internal versions.
+- We would reduce the need for "unsafe" conversions (a small security win) since
+  most APIs only have a single v1 version and so the entire request handling process
+  would simply use that version.
+- We would introduce more conversion when API request version differs from the
+  hub version. But beta APIs are off-by-default and we expect a lot less mixed
+  version API usage than in the past.
+
+This "hub version" change feels like something that could be made somewhat
+independent of this KEP.
+
+
+##### Mitigation: Avoid Conversion to Internal Type
+
+Requests are received as the versioned type, so it should be feasible to avoid
```

Comment:
It's not clear to me when defaulting would happen if this was done.  It's also not clear to me how we ensure that validation for different versions stays in sync